src/main/kotlin/com/chattriggers/ctjs/minecraft/libs/renderer/Font.kt

```diff
@@ -0,0 +1,397 @@
+        val utilityGraphics = (utilityImage.graphics as Graphics2D).also { it.font = font }
+
+        val fontMetrics = utilityGraphics.fontMetrics
+
+        for (index in characterData.indices) {
+            val character = index.toChar()
+            val characterBounds = fontMetrics.getStringBounds(character.toString() + "", utilityGraphics)
+
+            val width = characterBounds.width.toFloat() + 2 * MARGIN
+            val height = characterBounds.height.toFloat()
+
+            val characterImage = BufferedImage(
+                MathHelper.ceiling_double_int(width.toDouble()),
+                MathHelper.ceiling_double_int(height.toDouble()),
+                BufferedImage.TYPE_INT_ARGB
+            )
+
+            val graphics = (characterImage.graphics as Graphics2D).also {
+                it.font = font
+                it.color = Color(255, 255, 255, 0)
+                it.fillRect(0, 0, characterImage.width, characterImage.height)
+                it.color = Color.WHITE
+            }
+
+            graphics.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON)
+            graphics.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
+            graphics.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY)
+            graphics.setRenderingHint(RenderingHints.KEY_FRACTIONALMETRICS, RenderingHints.VALUE_FRACTIONALMETRICS_ON)
+
+            graphics.drawString(character.toString() + "", MARGIN, fontMetrics.ascent)
+
+            val textureId = GlStateManager.generateTexture()
+            createTexture(textureId, characterImage)
+
+            characterData[index] = CharacterData(
+                characterImage.width.toFloat(),
+                characterImage.height.toFloat(),
+                textureId
+            )
+        }
+
+        return characterData
+    }
+
+    private fun createTexture(textureId: Int, image: BufferedImage) {
+        val pixels = IntArray(image.width * image.height)
+
+        image.getRGB(0, 0, image.width, image.height, pixels, 0, image.width)
+
+        val buffer = BufferUtils.createByteBuffer(image.width * image.height * 4)
+        for (y in 0 until image.height) {
+            for (x in 0 until image.width) {
+                val pixel = pixels[y * image.width + x]
+
+                buffer.put((pixel shr 16 and 0xFF).toByte())
+                buffer.put((pixel shr 8 and 0xFF).toByte())
+                buffer.put((pixel and 0xFF).toByte())
+                buffer.put((pixel shr 24 and 0xFF).toByte())
+            }
+        }
+        (buffer as Buffer).flip()
+
+        GlStateManager.bindTexture(textureId)
+        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_NEAREST)
+        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_NEAREST)
+
+        GL11.glTexImage2D(
```

Comment:
The majority of this function can be eliminated. By using an internal format of `GL_RGBA8` (`GL_RBGA` would work too, but sized formats are preferred these days), a format of `GL_BGRA` and a type of `GL_UNSIGNED_INT_8_8_8_8_REV`, you can directly pass the `pixels` array to opengl, without any need for transformation.